src/wallet/orchard.h

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2021 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php .
+
+#ifndef ZCASH_WALLET_ORCHARD_H
+#define ZCASH_WALLET_ORCHARD_H
+
+#include <array>
+
+#include "primitives/transaction.h"
+#include "rust/orchard/keys.h"
+#include "rust/orchard/wallet.h"
+#include "zcash/address/orchard.hpp"
+
+class OrchardNoteMetadata
+{
+private:
+    OrchardOutPoint op;
+    libzcash::OrchardRawAddress address;
+    CAmount noteValue;
+    std::array<uint8_t, ZC_MEMO_SIZE> memo;
+    int confirmations;
+public:
+    OrchardNoteMetadata(
+        OrchardOutPoint op,
+        const libzcash::OrchardRawAddress& address,
+        CAmount noteValue,
+        const std::array<unsigned char, ZC_MEMO_SIZE>& memo):
+        op(op), address(address), noteValue(noteValue), memo(memo), confirmations(0) {}
+
+    const OrchardOutPoint& GetOutPoint() const {
+        return op;
+    }
+
+    void SetConfirmations(int c) {
+        confirmations = c;
+    }
+
+    int GetConfirmations() const {
+        return confirmations;
+    }
+
+    CAmount GetNoteValue() const {
+        return noteValue;
+    }
+};
+
+class OrchardWallet
+{
+private:
+    std::unique_ptr<OrchardWalletPtr, decltype(&orchard_wallet_free)> inner;
+
+public:
+    OrchardWallet() : inner(orchard_wallet_new(), orchard_wallet_free) {}
+    OrchardWallet(OrchardWallet&& wallet_data) : inner(std::move(wallet_data.inner)) {}
+    OrchardWallet& operator=(OrchardWallet&& wallet)
+    {
+        if (this != &wallet) {
+            inner = std::move(wallet.inner);
+        }
+        return *this;
+    }
+
+    // OrchardWallet should never be copied
+    OrchardWallet(const OrchardWallet&) = delete;
+    OrchardWallet& operator=(const OrchardWallet&) = delete;
+
+    void CheckpointNoteCommitmentTree() {
+        orchard_wallet_checkpoint(inner.get());
+    }
+
+    bool RewindToLastCheckpoint() {
+        return orchard_wallet_rewind(inner.get());
+    }
+
+    /**
+     * Add notes that are decryptable with IVKs for which the wallet
+     * contains the full viewing key to the wallet.
+     */
+    bool AddNotesIfInvolvingMe(const CTransaction& tx) {
```

Comment:
If we make `orchard_wallet_add_notes_from_bundle` infallible then this becomes infallible.